gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/source/DatasetCleanerSource.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.data.management.retention.source;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Properties;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.typesafe.config.Config;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.gobblin.configuration.SourceState;
+import org.apache.gobblin.configuration.WorkUnitState;
+import org.apache.gobblin.runtime.retention.DatasetCleanerTask;
+import org.apache.gobblin.runtime.retention.DatasetCleanerTaskFactory;
+import org.apache.gobblin.runtime.task.TaskUtils;
+import org.apache.gobblin.source.Source;
+import org.apache.gobblin.source.extractor.Extractor;
+import org.apache.gobblin.source.workunit.WorkUnit;
+import org.apache.gobblin.util.ConfigUtils;
+
+
+/**
+ * This class generates workunits from the configuration for cleaning datasets using instances of the
+ * {@link DatasetCleanerTask}
+ */
+@Slf4j
+public class DatasetCleanerSource implements Source<Object, Object> {
+  public static final String DATASET_CLEANER_SOURCE_PREFIX = "datasetCleanerSource";
+
+  /**
+   * This config holds a list of configuration names. Each configuration name defines a job that can have scoped config.
+   *
+   * So the list "config1, config2" means that two jobs are configured.
+   *
+   * Then the config can be scoped like:
+   * datasetCleanerSource.config1.state.store.db.table=state_table1
```

Comment:
Can we use a `DatasetFinder` to generalize the process of finding datasets to clean? For this case, we can create a `ConfigBasedDatasetFinder`. It makes the `DatasetCleanerSource` more general. @jack-moseley has done similar work for limited retention.

Reply:
This is not at the dataset level, but instead is configuration the connection to the state store to clean. A state store stores states for many datasets. For the current use cases of this a config store is not used. Will revisit this later if required.